src/targets/gpu/hip_gemm_impl.cpp

```diff
@@ -633,9 +632,19 @@ int32_t hip_gemm_finalize(context& ctx,
                           float beta,
                           int32_t solution_idx)
 {
-    auto gemm_item   = hip_gemm_impl(output_shape, input_shapes, alpha, beta);
-    int32_t solution = gemm_item.tune(ctx, input_shapes);
-    hip_gemm_save_solution(ctx, output_shape, input_shapes, solution_idx);
+    int solution;
+    auto gemm_item = hip_gemm_impl(output_shape, input_shapes, alpha, beta);
+    if(solution_idx == 0)
+    {
+        solution = gemm_item.tune(ctx, input_shapes);
+        hip_gemm_save_solution(ctx, output_shape, input_shapes, solution);
+    }
+    // If a tuned solution index is already given, don't tune again but validate
+    // in case the data was tuned with a different hipBLAS version.
+    else
+    {
+        solution = gemm_item.validate(ctx, input_shapes, solution_idx);
```

Comment:
This defaults the solution if it doesn't validate.  If the given index was tuned but doesn't validate, why wouldn't we want to tune again?

Reply:
> If the given index was tuned but doesn't validate, why wouldn't we want to tune again?

This is mainly for loading a compiled model from a file as the solution may not be valid with a different version of hipblaslt. Since we are loading the model it should happen somewhat quickly, so we should not be doing benchmarking.

Reply:
> This defaults the solution if it doesn't validate. If the given index was tuned but doesn't validate, why wouldn't we want to tune again?

@bpickrel Just to clarify, did you mean something like this?
```
    // If a tuned solution index is already given, don't tune again but validate
    // in case the data was tuned with a different hipBLASLt version.
    if(solution_idx != 0)
    {
        solution_idx = gemm_item.validate(ctx, input_shapes, solution_idx);
    }
    if(solution_idx == 0)
    {
        solution_idx = gemm_item.tune(ctx, input_shapes);
        hip_gemm_save_solution(ctx, output_shape, input_shapes, solution_idx);
    }
```

Reply:
Yes

Reply:
We shouldnt retune if the solution is invalid.